File: app/src/main/java/org/schabi/newpipe/local/playlist/RemotePlaylistManager.java
package org.schabi.newpipe.local.playlist;

import org.schabi.newpipe.database.AppDatabase;
import org.schabi.newpipe.database.playlist.dao.PlaylistRemoteDAO;
import org.schabi.newpipe.database.playlist.model.PlaylistRemoteEntity;
import org.schabi.newpipe.extractor.playlist.PlaylistInfo;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class RemotePlaylistManager {

    private final AppDatabase database;
    private final PlaylistRemoteDAO playlistRemoteTable;

    public RemotePlaylistManager(final AppDatabase db) {
        database = db;
        playlistRemoteTable = db.playlistRemoteDAO();
    }

    public Flowable<List<PlaylistRemoteEntity>> getPlaylists() {
        return playlistRemoteTable.getPlaylists().subscribeOn(Schedulers.io());
    }

    public Flowable<List<PlaylistRemoteEntity>> getPlaylist(final long playlistId) {
        return playlistRemoteTable.getPlaylist(playlistId).subscribeOn(Schedulers.io());
    }

    public Flowable<List<PlaylistRemoteEntity>> getPlaylist(final PlaylistInfo info) {
        return playlistRemoteTable.getPlaylist(info.getServiceId(), info.getUrl())
                .subscribeOn(Schedulers.io());
    }

    public Single<Integer> deletePlaylist(final long playlistId) {
        return Single.fromCallable(() -> playlistRemoteTable.deletePlaylist(playlistId))
                .subscribeOn(Schedulers.io());
    }

    public Completable updatePlaylists(final List<PlaylistRemoteEntity> updateItems,
                                       final List<Long> deletedItems) {
        return Completable.fromRunnable(() -> database.runInTransaction(() -> {
            for (final Long uid: deletedItems) {
                playlistRemoteTable.deletePlaylist(uid);
            }
            for (final PlaylistRemoteEntity item: updateItems) {
                playlistRemoteTable.upsert(item);
            }
        })).subscribeOn(Schedulers.io());
    }

    public Single<Long> onBookmark(final PlaylistInfo playlistInfo) {
        return Single.fromCallable(() -> {
            final PlaylistRemoteEntity playlist = new PlaylistRemoteEntity(playlistInfo);
            return playlistRemoteTable.upsert(playlist);
        }).subscribeOn(Schedulers.io());
    }

    public Single<Integer> onUpdate(final long playlistId, final PlaylistInfo playlistInfo) {
        return Single.fromCallable(() -> {
            final PlaylistRemoteEntity playlist = new PlaylistRemoteEntity(playlistInfo);
            playlist.setUid(playlistId);
            return playlistRemoteTable.update(playlist);
        }).subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/org/schabi/newpipe/local/playlist/RemotePlaylistManager.kt
/*
 * SPDX-FileCopyrightText: 2018-2025 NewPipe contributors <https://newpipe.net>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package org.schabi.newpipe.local.playlist

import androidx.room.withTransaction
import kotlinx.coroutines.flow.Flow
import org.schabi.newpipe.database.AppDatabase
import org.schabi.newpipe.database.playlist.model.PlaylistRemoteEntity
import org.schabi.newpipe.extractor.playlist.PlaylistInfo

class RemotePlaylistManager(private val database: AppDatabase) {
    private val playlistRemoteTable = database.playlistRemoteDAO()

    val playlists: Flow<List<PlaylistRemoteEntity>>
        get() = playlistRemoteTable.getPlaylists()

    fun getPlaylist(playlistId: Long): Flow<PlaylistRemoteEntity> {
        return playlistRemoteTable.getPlaylist(playlistId)
    }

    fun getPlaylist(info: PlaylistInfo): Flow<List<PlaylistRemoteEntity>> {
        return playlistRemoteTable.getPlaylist(info.serviceId.toLong(), info.url)
    }

    suspend fun deletePlaylist(playlistId: Long): Int {
        return playlistRemoteTable.deletePlaylist(playlistId)
    }

    suspend fun updatePlaylists(
        updateItems: List<PlaylistRemoteEntity>,
        deletedItems: List<Long>
    ) {
        database.withTransaction {
            deletedItems.forEach { playlistRemoteTable.deletePlaylist(it) }
            updateItems.forEach { playlistRemoteTable.upsert(it) }
        }
    }

    suspend fun onBookmark(playlistInfo: PlaylistInfo): Long {
        val playlist = PlaylistRemoteEntity(playlistInfo)
        return playlistRemoteTable.upsert(playlist)
    }

    suspend fun onUpdate(playlistId: Long, playlistInfo: PlaylistInfo): Int {
        val playlist = PlaylistRemoteEntity(playlistInfo).apply { uid = playlistId }
        return playlistRemoteTable.update(playlist)
    }
}
